### ISSUE TITLE
Deadlock when mixing Audio and Video

### ISSUE THREAD

**Issue by SteveMcFarlin** [centricular/gstcefsrc#69]
My issue is very similar to https://github.com/centricular/gstcefsrc/issues/59. The difference is my state flow is different.

gstreamer 1.20.3
gstcefsrc `master` with this patch applied: https://github.com/philn/gstcefsrc/commit/3b9d7eb6f5a576ae948cff92175c2f0ac2c1863d

My coded pipeline is identical to the one in the README. I enter into PAUSED to preload a web page. After it has loaded I enter into PLAYING. This will cause a deadlock when trying to send an EOS on the message bus. If I remove the audio then everything works as expected. Also replacing gstcefsrc/demux with a simple audio and video test source works fine. I did try setting queues to leaky, but this ends up causing AV sync issues.

My only thoughts are to maybe dynamically configure the output to the pipeline prior to going into PLAYING state, or modifying gstcefsrc to monitor the states and simply dump all AV data unless in PLAYING state. I really need the page to be preloaded and data flowing but not recording while in the PAUSED state. Maybe using a splitmuxsink might work.  I am totally open to any ideas anyone may have.

**Comment by SteveMcFarlin** [comment]
Adding a state transition monitor to cefsrc element class, and then adding a simple toggle to dump data unless in the PLAYING state fixes my issue.

**Comment by fairbairn** [comment]
Was this merged in as a patch or commit?

**Comment by MathieuDuponchelle** [comment]
@philn 's branch rebased with no conflicts on `master`, I have merged it now. You probably want to give `master` a spin again. As for this issue here, a test case with that two-step approach to going to PLAYING, showing the deadlock reliably would be helpful :)

**Comment by fairbairn** [comment]
That's great.  

I do want to add, that from master, we altered the pipeline to instead sink with an flvmux and route the output to an NGINX RTMP server, and it never hung.  We overtaxed the CPU, ran 12 pipelines in parallel on a GPU machine, and it never hung.

So I believe the problems stated before, although certainly a concern in terms of stability, may have been caused by the mux's characteristics.
